File: backend/app/core/monitoring.py
import logging

from app.core.config import Settings

ERROR_MONITORING_ENABLED_STATE_KEY = "error_monitoring_enabled"
monitoring_logger = logging.getLogger("projectops.monitoring")


def is_error_monitoring_configured(settings: Settings) -> bool:
    return settings.enable_error_monitoring and bool(settings.sentry_dsn.strip())


def configure_error_monitoring(settings: Settings) -> bool:
    if not is_error_monitoring_configured(settings):
        return False

    import sentry_sdk
    from sentry_sdk.integrations.starlette import StarletteIntegration

    sentry_sdk.init(
        dsn=settings.sentry_dsn,
        environment=settings.sentry_environment.strip() or settings.environment,
        integrations=[StarletteIntegration()],
        send_default_pii=False,
        before_send=sanitize_monitoring_event,
    )
    return True


def capture_unhandled_exception(request, error: Exception) -> None:
    if not getattr(request.app.state, ERROR_MONITORING_ENABLED_STATE_KEY, False):
        return

    try:
        import sentry_sdk

        request_id = getattr(request.state, "request_id", None)
        if request_id:
            sentry_sdk.set_tag("request_id", request_id)
            sentry_sdk.set_context(
                "projectops",
                {
                    "request_id": request_id,
                    "path": request.url.path,
                    "method": request.method,
                },
            )
        sentry_sdk.capture_exception(error)
    except Exception:
        monitoring_logger.warning("error_monitoring_capture_failed", exc_info=True)


def sanitize_monitoring_event(event: dict, hint: dict | None = None) -> dict:
    request = event.get("request")
    if isinstance(request, dict):
        request.pop("data", None)
        request.pop("cookies", None)
        headers = request.get("headers")
        if isinstance(headers, dict):
            request["headers"] = {
                key: value
                for key, value in headers.items()
                if key.lower() not in {"authorization", "cookie", "set-cookie"}
            }
        url = request.get("url")
        if isinstance(url, str):
            request["url"] = url.split("?", 1)[0]
        query_string = request.get("query_string")
        if query_string:
            request["query_string"] = "[Filtered]"
    return event

File: frontend/src/observability/monitoring.ts
import * as Sentry from "@sentry/react";

type MonitoringEnvironment = {
  VITE_ENABLE_ERROR_MONITORING?: string;
  VITE_SENTRY_DSN?: string;
  VITE_SENTRY_ENVIRONMENT?: string;
  MODE?: string;
};

type CaptureContext = {
  requestId?: string;
};

let monitoringEnabled = false;

export function isFrontendMonitoringConfigured(env: MonitoringEnvironment = import.meta.env): boolean {
  return env.VITE_ENABLE_ERROR_MONITORING === "true" && Boolean(env.VITE_SENTRY_DSN?.trim());
}

export function initFrontendMonitoring(env: MonitoringEnvironment = import.meta.env): boolean {
  if (!isFrontendMonitoringConfigured(env)) {
    monitoringEnabled = false;
    return false;
  }

  Sentry.init({
    dsn: env.VITE_SENTRY_DSN?.trim(),
    environment: env.VITE_SENTRY_ENVIRONMENT?.trim() || env.MODE || "local",
    sendDefaultPii: false,
    beforeSend: sanitizeFrontendEvent,
  });
  monitoringEnabled = true;
  return true;
}

export function captureFrontendException(error: unknown, context: CaptureContext = {}) {
  if (!monitoringEnabled) return;

  if (context.requestId) {
    Sentry.setTag("request_id", context.requestId);
    Sentry.setContext("projectops", { request_id: context.requestId });
  }
  Sentry.captureException(error);
}

function sanitizeFrontendEvent(event: Sentry.ErrorEvent): Sentry.ErrorEvent {
  const request = event.request;
  if (request) {
    delete request.cookies;
    delete request.data;
    if (request.headers) {
      Object.keys(request.headers).forEach((key) => {
        if (["authorization", "cookie", "set-cookie"].includes(key.toLowerCase())) {
          delete request.headers?.[key];
        }
      });
    }
    if (request.url) {
      request.url = request.url.split("?", 1)[0];
    }
    if (request.query_string) {
      request.query_string = "[Filtered]";
    }
  }
  return event;
}
